### PR TITLE
Update checkout calculations for promotion

### PR DESCRIPTION
**Pull request by IKarbowiak**
Creates checkout discounted objects for promotions.

- Add relation to `promotion rule` for `OrderDiscount` and `CheckoutLineDiscount`
- Update `CheckoutLineDiscount` creation - assign applied promotion rules
- Update variant price calculation - use the `discounted_price` from `channel_listing`. When the custom price is provided, apply the valid rules on the custom price.
- Use the `discounted_price` from variant channel listings for checkout calculations

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are either absent or guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migrations are either absent or optimized for zero downtime
* [ ] The changes are covered by test cases

### CODE REVIEW

File: saleor/checkout/calculations.py [hunk 1]
```diff
@@ -243,6 +246,8 @@ def _fetch_checkout_prices_if_expired(
     charge_taxes = get_charge_taxes_for_checkout(checkout_info, lines)
     should_charge_tax = charge_taxes and not checkout.tax_exemption
 
+    create_or_update_discount_objects_from_promotion_for_checkout(lines)
+    # TODO: to remove promotion CRUD
```
Comment (fowczarek) [inline saleor/checkout/calculations.py:250]: Will it be fixed in a separate PR?

File: saleor/order/utils.py [hunk 2]
```diff
@@ -224,22 +224,15 @@ def create_order_line(
...
     # vouchers are not applied for new lines in unconfirmed/draft orders
     untaxed_unit_price = variant.get_price(
-        product,
-        collections,
-        channel,
         channel_listing,
-        discounts,
         price_override=price_override,
     )
+    # TODO: rewrite for promotion
```
Comment (fowczarek) [inline saleor/order/utils.py:230]: Will it be fixed in a separate PR?

**Review by fowczarek — APPROVED** [review]

File: saleor/checkout/calculations.py (hunk 1 above)
Comment (IKarbowiak) [reply saleor/checkout/calculations.py:250]: Yes

File: saleor/order/utils.py [hunk 3]
```diff
@@ -285,6 +278,7 @@ def create_order_line(
     )
 
     unit_discount = line.undiscounted_unit_price - line.unit_price
+    # TODO: rewrite for promotions
```
[the remark this answers is not in the document]
Comment (IKarbowiak) [reply saleor/order/utils.py:281]: Yes

File: saleor/order/utils.py (hunk 2 above)
Comment (IKarbowiak) [reply saleor/order/utils.py:230]: Yes

File: saleor/checkout/fetch.py [hunk 4]
```diff
@@ -65,6 +63,13 @@ def get_sale_discount(self) -> Optional["CheckoutLineDiscount"]:
                 return discount
         return None
 
+    def get_promotion_discounts(self) -> List["CheckoutLineDiscount"]:
+        discounts = []
+        for discount in self.discounts:
+            if discount.type == DiscountType.PROMOTION:
+                discounts.append(discount)
+        return discounts
```
Comment (tomaszszymanski129) [inline saleor/checkout/fetch.py:71]:
```suggestion
        return filter(lambda discount: discount.type == DiscountType.PROMOTION, self.discounts)
```

File: saleor/product/models.py [hunk 5]
```diff
@@ -647,22 +647,20 @@ def get_base_price(
...
-            variant_id=self.id,
+        """Return the variant discounted price with applied promotions.
+
+        If a custom price is provided, return the price with applied discounts from
+        valid promotion rules for this variant.
+        """
+        if price_override is None:
+            return channel_listing.discounted_price or channel_listing.price
+        price: "Money" = self.get_base_price(channel_listing, price_override)
+        rules = channel_listing.promotion_rules.all()
```
[the remark this answers is not in the document]
Comment (tomaszszymanski129) [reply saleor/product/models.py:661]: I agree about the queries, it could create an N+1 query issue when someone uses that within resolver.

**Review by fowczarek — APPROVED** [review]

**Review by korycins — APPROVED** [review]
